### PR TITLE
Posts.com : Milestone : Integration tests and improving quires performance

### PR DESCRIPTION
**Pull request by ahmed025566**
Hello there, code reviewer.
Hope you are doing well.

In this milestone you could see that I have done the following:

- I have written test cases for both user and posts templates so that it can simulates the user interaction with the site.
- I have improved quires performance by reducing the number of fetching requests every time the user request the posts as 
  it is explained in the attached screenshots.
post index before

post index after
post show before
post show after

### CODE REVIEW

File: spec/integration/post_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  it 'should display how many comments a post has' do
+    visit user_posts_path(@user)
+    expect(page).to have_content("Comments: #{@post1.CommentsCounter}")
+  end
+
+  it 'should display how many likes a post has' do
+    visit user_posts_path(@user)
+    expect(page).to have_content("Likes: #{@post1.LikesCounter}")
+  end
+
```
Comment (topeogunleye) [inline spec/integration/post_index_spec.rb:68]: 🚀 **Code Review on Feature Test for User Post Index Page** 🚀

Hi there! Fantastic effort on your feature tests for the User Post Index page. You're on the right track! Here's the feedback:

✅ **Positive Aspects:**
- **Comprehensive**: You've comprehensively addressed the required assertions! It's great to see checks for elements like the user's profile picture, username, post counts, and interactions like comments and likes. 🎉
- **Well-Organized Tests**: The tests are neatly organized, enhancing readability and ease of maintenance. 📚
- **Navigation Testing**: Incorporating tests for navigation by checking redirects when clicking on a post is spot-on. This is crucial for a good user experience. 👌

🛑 **Required Change:**
- **Pagination Testing**: The project specification requires a test for pagination, but it seems to be missing. Please add a test to ensure that pagination controls appear when there are more posts than can fit on a single page. This is crucial for user experience and application performance. 📃⏭️

🔄 **Optional Changes:**
- **Ruby Naming Conventions**: While not critical, adhering to Ruby's naming conventions would be beneficial. This includes using snake_case for variable and method names. 🐍
- **Testing for "First" Comments**: The current test for comments checks all comments instead of just the "first" as implied by the test description. Aligning the description with the test content or adjusting the test to match its description would be beneficial for clarity. 🔍
- **Data Creation**: Consider using factories for setting up test data to make your tests even cleaner and potentially more performant. This can also help with scalability as your test suite grows. 🏗️
- **Test Suite Performance**: Grouping common setup with `context` blocks and using `let` for lazy loading can improve test speed, something to keep in mind as your test suite expands. 🚅

Keep up the fantastic work, and always strive for tests that reflect real user behavior as closely as possible. Happy coding! 🌟💻

File: spec/integration/post_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    visit user_post_path(user_id: @user.id, id: @post.id)
+    expect(page).to have_content(@post.Title)
+  end
+
+  it 'should display the author name' do
+    visit user_post_path(user_id: @user.id, id: @post.id)
+    expect(page).to have_content(@user.Name)
+  end
+
+  it 'should display the comments counter' do
```
Comment (topeogunleye) [inline spec/integration/post_show_spec.rb:20]: 🌟 **Code Review on Feature Test for User Post Show Page** 🌟

Hey there! You've done a commendable job on the feature tests for the User Post Show page. Your effort in covering the necessary assertions is greatly appreciated. Here's the breakdown:

✅ **Positive Feedback:**
- **Test Coverage**: You've successfully covered all the required assertions as specified in the project requirements. This is great for ensuring each aspect of the post show page is functioning as expected. 📊
- **Clarity**: The tests are clear and to the point, reflecting a good understanding of both the project requirements and RSpec testing framework. 🎯
- **Code Consistency**: The usage of instance variables for setup and calling attributes in expectations provides a consistent pattern that enhances the readability of your tests. 👍

🔄 **Suggestions for Improvement:**
- **Validation of Counter Properties**: You're testing the presence of comments and likes counters, which is excellent. However, ensure the counters are not just present but also displaying the correct numbers. This may require additional setup in your `before` block to create likes and comments and verify that the counters reflect these creations accurately. 🔢
- **Edge Cases [OPTIONAL]**: Consider adding tests for edge cases, such as when there are no comments or likes. This helps ensure the page behaves correctly in these scenarios as well. 🛡️

🛠 **Required Change:**
- **Likes Counter Check**: I noticed that you are testing for both the comments counter and likes counter in a single `it` block. While this might seem efficient, each `it` block should ideally test one and only one aspect to keep the tests atomic. This will also help to identify the cause of a test failure more quickly. Please separate these into individual tests. 🧪

👏 Overall, you're doing fantastic work! Keep testing thoroughly and ensuring your test scenarios are as close to real-world usage as possible. Happy coding! 🌟💼

File: app/controllers/posts_controller.rb [hunk 3]
```diff
@@ -2,10 +2,12 @@ class PostsController < ApplicationController
   def index
     @user = User.find(params[:user_id])
```
Comment (topeogunleye) [inline app/controllers/posts_controller.rb:3]: 🔍 **Code Review for PostsController Index Action** 🔍

Hello! It looks like you've made some progress with the PostsController. However, there is a specific requirement regarding the N+1 problem that needs to be addressed. Let's take a look:

🚀 **Current Implementation:**
```ruby
def index
  @user = User.find(params[:user_id])
  @posts = @user.posts
  @comments = Comment.all
  @users = User.all
end
```

🔧 **Issue with N+1 Queries:**
- In the current implementation, you are retrieving all posts for a user with `@posts = @user.posts`, which can lead to N+1 query problems when you try to access comments for each post. This is because it will fire a separate query for each post's comments when accessed in the view, thus reducing the efficiency of data retrieval.

🎯 **Required Change:**
- To solve this problem and meet the project requirements, you need to preload the associated comments with the posts. You can achieve this by using `.includes(:comments)` method. This will load all the related comments for the posts in a single query, hence optimizing the database calls.

📝 **Suggested Implementation:**
```ruby
def index
  @user = User.find(params[:user_id])
  @posts = @user.posts.includes(:comments)
  # The following line may also need optimization depending on further requirements.
  @comments = Comment.all
  @users = User.all
end
```

📊 **Before and After Fix:**
- As per the requirements, make sure you provide screenshots of your logs before and after the fix to illustrate that the N+1 query issue has been resolved. Remember not to use the bullet gem for this purpose as per the instructions.

Keep up the good work, and once this change is implemented, you'll be right on track with the project requirements! 👍

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi @Gh-Reza, @ahmed025566,

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

Try-Again

> ## (Highlights) Good Points: 💪 

1. **Capybara Integration 🧪**:
   - You've correctly included the Capybara gem in your Gemfile under the appropriate groups. This ensures that your integration tests can simulate how a user interacts with your application. 👏

2. **User Index Page Tests 📄**:
   - Your spec file is focused and solely tests the User index page as required. Breaking down the tests to address each aspect has made your tests thorough and clear. 🎯
   - The tests confirm visibility of usernames, profile pictures, and post counts for all users, which aligns perfectly with the requirements. This is crucial for a positive user experience on the index page. ✅
   - The redirection test to a user's show page upon clicking their profile is a nice touch to ensure the navigation flow is intuitive and functional. 🔗

3. **User Show Page Tests 🖼️**:
   - For the User show page, your tests comprehensively assert the presence of the user's profile picture, username, and post count, which is fantastic for detailed user profiles. 🌐
   - Verifying the user's bio and the first 3 posts provides a sneak peek into the user's activity, meeting the requirement for partial content display. 📚
   - The "view all posts" button test confirms that users can explore more content if they wish, which is great for engagement. 🚀
   - Your redirect tests ensure that the user can navigate to a post's show page and the user's post index page, solidifying the user journey within the application. 👣

Overall, you've done a commendable job of creating focused, requirement-driven tests that validate crucial functionalities of the user interface. This attention to detail enhances the reliability and user-friendliness of the application. Keep it up! 🌈👍

## Required Changes ♻️
_Please Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
_Please Check the comments under the review._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @topeogunleye in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Shedrack-Sunday — APPROVED** [review]
Hi @ahmed025566        👍 

Your project is complete! There is nothing else to say other than... it's time to merge it :shipit:
Congratulations! 🎉

image


Highlights
- [x] Good project organization ✔️
- [x] Nice use of Rails testing features  ✔️
- [x] Kudus on fixing all changes suggested by the last reviewer   
- [x] Kudus on meeting the requirements   👏 


Optional suggestions
Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application.


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).
